File: leetcode/easy/MajorityElement169.java
package easy;

// https://leetcode.com/problems/majority-element/
// 169. Majority Element
// https://www.youtube.com/watch?v=AoX3BPWNnoE

public class MajorityElement169 {
    public static void main(String[] args) {
        System.out.println(new MajorityElement169().majorityElement(new int[] { 3, 2, 3 }));
    }

    // 2,2,1,1 | 1,2 | 2
    // Moore voting algorithm
    // the idea is to remove minority elements using counts
    public int majorityElement(int[] nums) {
        int count = 0;
        int candidate = 0;
        for (int num : nums) {
            if (count == 0) {
                candidate = num;
            }
            if (candidate == num) {
                count++;
            } else {
                count--;
            }
        }
        return candidate;
    }
}

File: leetcode/easy/MajorityElement169.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:

        # O(nlogn)
        # nums.sort()
        # return nums[len(nums)//2]

        # O(n)
        # freq_map = {}
        # res, max_count = 0, 0

        # for n in nums:
        #     freq_map[n] = 1 + freq_map.get(n, 0)
        #     res = n if freq_map[n] > max_count else res
        #     max_count = max(freq_map[n], max_count)
        # return res

        # Moore Voting algorithm
        # Boyer-Moore algorithm

        res, count = 0, 0
        for n in nums:
            if count == 0:
                res = n
            count += (1 if n == res else -1)

        return res

